### PR TITLE
Fix dark mode for metadata UI

### PR DESCRIPTION
**Pull request by ajbozarth**
This is a follow up to #1327 fixing various bugs as well as adding
dark mode support.

Dark Mode support for material-ui has now been abstracted to a new
ui-component and MetadataEditor has been updated to use it.

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by ajbozarth** [comment]
Note those those I tagged for review, this PR touches a lot of code from #1293 if you could make sure I haven't broken anything from that PR (I spent all afternoon making sure I didn't but odd are I missed something)

**Comment by ajbozarth** [comment]
For reasons I am still unsure of, this PR makes the code editor in the MetadataEditor (used by code snippets) not load. This is (obviously) causing the tests to fail. I will look into why this is happening tomorrow

File: packages/ui-components/src/JpThemeProvider.tsx
Comment (bourdakos1) [inline packages/ui-components/src/JpThemeProvider.tsx:64]: This should go into an effect hook

**Review by bourdakos1 — CHANGES_REQUESTED** [review]

File: packages/ui-components/src/JpThemeProvider.tsx
Comment (ajbozarth) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: like this? 
```
  useEffect(() => {
    if (props.themeManager) {
      props.themeManager.themeChanged.connect(updateTheme);
    }
  });
```
I just googled it and it seems to the the function equivalent of componentDidMount, why should this be in an effect hook?

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: you also need a dependency array otherwise this will run every render

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]:
```
useEffect(() => {
    if (themeManager) {
     themeManager.themeChanged.connect(updateTheme);
    }
  }, [themeManager]);
```

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: this will make it so it only connects whenever themeManager or updateTheme change, does this have a disconnect function? because we should do that here as well

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: I'll write up a quick demo with notes on how this works with hooks

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]:
```js
// The themes should be created outside of the component, otherwise they will be re-created
// every render which could be expensive.
const lightTheme = createMuiTheme({
  palette: {
    type: 'light'
  }
});

const darkTheme = createMuiTheme({
  palette: {
    type: 'dark'
  }
});

function isLightTheme(themeManager) {
  return themeManager.theme && themeManager.isLight(themeManager.theme);
}

function JpThemeProvider({ themeManager, children, updateWidget }) {
  // variables in a component should not be directly modified by side effects. If we want to 
  // modify a variable from an effect/callbacks it should be a state hook.
  // Note: It is normally frowned upon to derive state from props, because the can become 
  // stale. In this case it is okay because we update this value through an effect.
  const [isLight, setIsLight] = useState(isLightTheme(themeManager))

  useEffect(() => {
    // Any type of listeners are side effects and should be set up and torn down inside
    // an effect hook. Otherwise, you will register a listener every render and end up
    // a big leak

    // `updateTheme` is small and isn't called anywhere else so we can define it inside
    // of the effect so we don't need to add it to the dependency array.
    // It's important to note that if this is defined inside of the component and listed in
    // the dependency array the effect would be rerun on every render because a new 
    // function is created every render. You can avoid this by using a `callback` hook
    // which will memoize the function when possible 
    function updateTheme() {
      setIsLight(isLightTheme(themeManager))
      updateWidget()
    }

    if (themeManager) {
      themeManager.themeChanged.connect(updateTheme);
    }
    return () => {
      // disconnect when the when the component is unmounted/the dependency array has 
      // changed and a new connection has been made.
      themeManager.themeChanged.disconnect(updateTheme);
    }
  // This is the dependency array, anytime one of these change this effect will be ran. It will
  // also run when the component is first mounted.
  // Notice, that `updateWidget` is listed here, but not `isLightTheme` or `setIsLight`. This
  // is because any `setState` functions are guaranteed not to change as well as the function
  // defined outside of this component. However, `updateWidget` is a prop and could change 
  // from render to render. (There is an eslint rule that will automatically generate this array for you)
  }, [themeManager, updateWidget])

  return (
    <ThemeProvider theme={isLight ? lightTheme : darkTheme}>
      {children}
    </ThemeProvider>
  )
}
```

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: Here is an example of `updateTheme` with the callback hook:
```
const updateTheme = useCallback(() => {
  setIsLight(isLightTheme(themeManager))
  updateWidget()
}, [themeManager, updateWidget])
```

Comment (bourdakos1) [reply packages/ui-components/src/JpThemeProvider.tsx:64]: there's also the option to memoize the theme and use it inside of the component:
```
const theme = useMemo(() => {
  return createMuiTheme({
    palette: {
      type: isLight ? 'light' : 'dark'
    }
  });
}, [isLight])
```

**Comment by ajbozarth** [comment]
I have made some changes to match @bourdakos1 comments above but I was dealing with rendering race conditions all afternoon. @bourdakos1 I'll ping you tomorrow and see if we can hop on a Webex and you can help me try to figure it out.

If I cant figure it out by early afternoon (pacific time) I’m going to partially roll back my work so that the part of the PR that fixes dark mode for the drop down can at least be merged since that part already works fine. That will unblock the 2.1 release, and then we can deal with moving the MetadataEditor to use the JpThemeProvider component in a follow up PR since that’s where the race condition is happening.

**Review by bourdakos1 — APPROVED** [review]
This looks good to me

**Comment by ajbozarth** [comment]
@lresende once tests pass this should be good to merge (unless you want more reviews first)

File: packages/ui-components/src/JpThemeProvider.tsx
Comment (lresende) [inline packages/ui-components/src/JpThemeProvider.tsx:32]: I would like to avoid introducing `Jp`, 'Lab`, or `Elyra`prefixes to class names or similar suffixes.

File: packages/ui-components/src/JpThemeProvider.tsx
Comment (lresende) [inline packages/ui-components/src/JpThemeProvider.tsx:38]: to promote usage, could we add some inline docs on what these do?

Comment (ajbozarth) [reply packages/ui-components/src/JpThemeProvider.tsx:38]: @bourdakos1 you understand this better, could you add an inline comment?

Comment (ajbozarth) [reply packages/ui-components/src/JpThemeProvider.tsx]: done

File: packages/ui-components/src/JpThemeProvider.tsx
Comment (ajbozarth) [reply packages/ui-components/src/JpThemeProvider.tsx]: done

**Review by lresende — APPROVED** [review]

**Pull request merged**
